### ISSUE TITLE
dream.entropy.standardize() results in NaN when standard deviation is zero

### ISSUE THREAD

**Issue by danaugrs** [bumps/bumps#21]
Please see [here](https://github.com/bumps/bumps/blob/master/bumps/dream/entropy.py#L72). `standardize()` is not considering the case when the computed standard deviation `sigma` is zero. When that happens, the output is `NaN` (or a tensor of `NaN`s). 

To fix this I recommend doing [the same as scikit-learn](https://github.com/scikit-learn/scikit-learn/blob/7389dbac82d362f296dc2746f10e43ffa1615660/sklearn/preprocessing/data.py#L70), that is, when `sigma` is 0, change it to 1. However it should be implemented such that it works both when `sigma` is a number and a tensor.

PS: Thanks for making this cool library!

**Comment by pkienzle** [comment]
z-score should be zero for all x if all x are identical.  I'm not certain that setting sigma to one is the best way to achieve that.  In particular, I would rather through an error than return an incorrect entropy estimate.

If you are working with continuous parameters, then x should never be identical: the partial of chi^2 with respect to parameter p_k should be finite, which should lead to some amount of variation in the sample set.  So having a sigma of zero on a parameter is an indication that the sampling has gone badly.  Assuming the distribution is really that narrow that the MCMC stepper can't sample from it then the entropy should be -inf.  You can see this by considering an uncorrelated multivariate normal with entropy H = log(det(2 pi e Sigma))/2 in the limit as Sigma goes to zero in one dimension.  By returning zero for sigma, the entropy calculation using sklearn_log_density should naturally return -inf.  Not sure if the kde will work correctly with X constant; probably depends on the algorithm.

Discrete data is different.  Here we need to use H = sum P(x) log P(x), where P(x) = #x/n.  In this case, with x identical, P(x) = 1 and P(x) log P(x) = 0.   Also, should probably choose a kde algorithm which knows about discrete distributions.

**Comment by danaugrs** [comment]
I'm calling `kde_entropy_sklearn` a very large number of times and some times it has happened that all samples are the same (without sampling having gone bad). I'd rather the entropy returned were `-Inf` in that case because that would work perfectly for me.

For context, I'm using this in a novel reinforcement learning algorithm.

**Comment by pkienzle** [comment]
I posted a fix, and it should now produce an entropy of -inf if all points are the same in any dimension.  I haven't yet cut a new version.  Probably a couple of months before it shows up on pypi.

I'm not particularly happy with this solution since it implies much more knowledge than we actually have (we are limited in our information by the sample size).  However, if this is good enough for your purposes, I'm happy to let it stand.  

I haven't tested what happens when the degeneracy doesn't lie on an axis.  Should be able to see this by rotating the covariance matrix diag([1,1,0]) and using it for a multivariate normal.

I still don't understand how a well-behaved sampler operating on continuous variables can yield a set of points with identical values in any dimension, except perhaps if the sample size is tiny.
